### ISSUE TITLE
UNet Implementation details

### ISSUE THREAD

**Issue by SebastienTs** [ELEKTRONN/elektronn3#52]
I noticed some subtle differences in the implementation of your unet and the one in this repo: 
https://github.com/johschmidt42/PyTorch-2D-3D-UNet-Tutorial. 

Basically, the batch normalization is performed after the ReLu layers instead of the convolutional layers.
Also, the weight are initialized by Xavier_uniform method instead of being set constant.

For my application it leads to better accuracy and speeds up the training by quite some margin.
Have been testing both approaches and do you have a strong opinion about it?

Related to this, I'm wondering if you may consider including new UNet variants, UNet with pre-trained decoder (at least for the 2D cases), or maybe some transformer based networks. 

If not, is there a repository that you would particularly recommend to import Pytorch models from for bioimage applications (3D fluorescence LM and 3D EM)?

**Comment by mdraw** [comment]
I have chosen this order of BN / ReLU layers because that was used in other implementations that I looked at (including the paper that originally proposed batch norm) and at that time I knew about optimized fused conv+bn+relu ops but had not seen conv+relu+bn ops.
But if you think about it I guess BN after ReLU makes more sense because ReLU after BN kind of defeats the purpose of normalization: inputs are zero mean, unit variance and ReLU just sets all negative elements to zero (i.e. discarding half of the input elements on average) and reduces variance by half. That doesn't really make sense so if you don't have to stay compatible with earlier neural network architectures I would indeed suggest to switch the order. I also just found a good reddit discussion about this: https://www.reddit.com/r/MachineLearning/comments/67gonq/d_batch_normalization_before_or_after_relu/

Regarding initialization, I would just do experiments with different initializations and use what works best for you. The optimal initialization can vary depending on the architecture.

I haven't done extensive tests of initialization and batch norm placement myself. I can only say that enabling batch norm makes trainings converge much faster consistently than disabling it, but that's not that surprising I guess :)

I am not planning to include more recent neural network architectures because in my experience almost none of the architectures that have been published since the U-Net / 3D U-Net paper actually work better than a well-adapted U-Net, despite of their own reported benchmark results.

If you are looking for well-written fancy new network architecture implementations, I suggest the collection in https://github.com/Project-MONAI/MONAI/tree/dev/monai/networks/nets. You can just install monai and import the network models from the monai package and use them in the same way as the elektronn3-included models without any further changes because they are also just `torch.nn.Module` subclasses.

**Comment by SebastienTs** [comment]
Thanks for elaborating on this topic!

Since this is the only other difference I could figure out between the unet
I'm testing (with inverted ReLU / BN) and yours,
do you confirm that the dummy attention layers (attention = False) play
absolutely no role as the name suggest?

It's probably an extreme case but in my application I'm trying to fill up
tubes (bright membrane, dark inside) and while the first unet converges
quite fast to a satisfying solution, I can't manage to obtain similar
results with the elektronn3 unet even with much longer training or
deeper/wider networks.

I also had the feeling that most unet variants do not bring much more in as
compared to a well configured/trained classical unet, at least for bioimage
analysis where, as opposed to natural scenes, whole objects tend to be
uniformely textured and a very large context is not necessarily required.

I'm unsure you can access it but this blog attempts to very schematically
summarize the differences between most unet variants:
https://link.medium.com/qc9T5MI5nCb

The exact application case is not described but the author also suggests
that the improvements are overall marginal over the vanilla U-Net, and that
Attention U-Net and U-Net3+ are the best performing.

ResUnet, U2Net and Transformer based U-Net required apparently much more
data and training to get similar performance. I believe that your attention
U-Net is similar to the architecture described in the blog but I'm not
fully positive on this.

I also heard from some colleagues that ResUnet brought substancial
improvements for a specific medical imaging problem they tackled and that
multiplanar U-Net (not discussed in this blog) is able to perform close to
3D U-Net at a lower complexity.

Then there are the newest comers such as Mask2former and OneFormer. Again,
I'm not sure if they could bring any disruptive gain for bioimage analysis,
but generically trained networks may already offer good performance without
training and they also tackle instance segmentation, which can be a big
plus.
